src/parquet/column_reader.cc

```diff
@@ -114,18 +114,51 @@ bool TypedColumnReader<TYPE>::ReadNewPage() {
       // Read a data page.
       num_buffered_values_ = current_page_header_.data_page_header.num_values;
 
-      // Read definition levels.
+      //OK for flat schema. Will be updated after schema resolution.
+      int max_repetition_level_ = 0;
+      int max_definition_level_ = 1;
+
+      //Data page Layout: Definition Levels - Repetition Levels - encoded values.
+      //Levels are encoded as rle-bit-packed-hybrid.
+      //TODO: make a level class to avoid code duplication.
       if (schema_->repetition_type != FieldRepetitionType::REQUIRED) {
-        int num_definition_bytes = *reinterpret_cast<const uint32_t*>(buffer);
-        buffer += sizeof(uint32_t);
-        definition_level_decoder_.reset(
-            new RleDecoder(buffer, num_definition_bytes, 1));
-        buffer += num_definition_bytes;
-        uncompressed_len -= sizeof(uint32_t);
-        uncompressed_len -= num_definition_bytes;
-      }
+        parquet::Encoding::type encoding;
+        //Read definition levels
+        if (max_definition_level_ > 0) {
+          int def_bit_width = BitUtil::Log2(max_definition_level_ + 1);
+          int num_definition_bytes = 0;
+          encoding  = current_page_header_.data_page_header.definition_level_encoding;
+          if (encoding == parquet::Encoding::RLE || encoding == parquet::Encoding::BIT_PACKED) {
+            num_definition_bytes = *reinterpret_cast<const uint32_t*>(buffer);
+            buffer += sizeof(uint32_t);
+            uncompressed_len -= sizeof(uint32_t);
+            definition_level_decoder_.reset(
+                new RleDecoder(buffer, num_definition_bytes, def_bit_width));
+            buffer += num_definition_bytes;
+            uncompressed_len -= num_definition_bytes;
+          } else {
+            throw ParquetException("Unknown encoding type for levels.");
+          }
+        }
 
-      // TODO: repetition levels
+         //Read repetition levels
+        if (max_repetition_level_ > 0) {
+          int rep_bit_width = BitUtil::Log2(max_repetition_level_ + 1);
+          int num_repetition_bytes = 0;
+          encoding  = current_page_header_.data_page_header.repetition_level_encoding;
+          if (encoding == parquet::Encoding::RLE || encoding == parquet::Encoding::BIT_PACKED) {
+            num_repetition_bytes = *reinterpret_cast<const uint32_t*>(buffer);
+            buffer += sizeof(uint32_t);
+            uncompressed_len -= sizeof(uint32_t);
+            repetition_level_decoder_.reset(
+	        new RleDecoder(buffer, num_repetition_bytes, rep_bit_width));
+            buffer += num_repetition_bytes;
+            uncompressed_len -= num_repetition_bytes;
+          } else {
+            throw ParquetException("Unknown encoding type for levels.");
```

Comment:
Can you refactor this to avoid code duplication? Perhaps combine this effort to create a patch for PARQUET-462 to kill two birds with one stone